src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,187 @@
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Is Async caller method</param>
+        /// <param name="ct">Cancellation token</param>
+        public async Task WriteShortAsync(short v, bool isAsync, CancellationToken ct)
+        {
+            await EnsureBufferSpaceAsync(2, isAsync, ct).ConfigureAwait(false);
+            BinaryPrimitives.WriteInt16LittleEndian(GetWritableSpan(2), v);
+            _outBytesUsed += 2;
+        }
+
+        /// <summary>
+        /// Writes int value to out buffer
+        /// </summary>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Is Async caller method</param>
+        /// <param name="ct">Cancellation token</param>
+        public async Task WriteIntAsync(int v, bool isAsync, CancellationToken ct)
+        {
+            await EnsureBufferSpaceAsync(4, isAsync, ct).ConfigureAwait(false);
+            BinaryPrimitives.WriteInt32LittleEndian(GetWritableSpan(4), v);
+            _outBytesUsed += 4;
+        }
+
+        /// <summary>
+        /// Writes unsigned short value to out buffer
+        /// </summary>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Is Async caller method</param>
+        /// <param name="ct">Cancellation token</param>
+        public Task WriteUnsignedShortAsync(ushort v, bool isAsync, CancellationToken ct) => WriteShortAsync((short)v, isAsync, ct);
+
+        /// <summary>
+        /// Writes unsigned int value to out buffer
+        /// </summary>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Is Async caller method</param>
+        /// <param name="ct">Cancellation token</param>
+        public Task WriteUnsignedIntAsync(uint v, bool isAsync, CancellationToken ct) => WriteIntAsync((int)v, isAsync, ct);
+
+        /// <summary>
+        /// Writes float value to out buffer
+        /// </summary>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Is Async caller method</param>
+        /// <param name="ct">Cancellation token</param>
+        public async Task WriteFloatAsync(float v, bool isAsync, CancellationToken ct)
+        {
+            if (float.IsInfinity(v) || float.IsNaN(v))
+            {
+                throw new ArgumentOutOfRangeException(nameof(v), "Float value is out of range.");
+            }
+            await EnsureBufferSpaceAsync(4, isAsync, ct).ConfigureAwait(false);
+            BinaryPrimitives.WriteInt32LittleEndian(GetWritableSpan(4), BitConverter.SingleToInt32Bits(v));
+            _outBytesUsed += 4;
+        }
+
+        /// <summary>
+        /// Writes long value to out buffer
+        /// </summary>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Is Async caller method</param>
+        /// <param name="ct">Cancellation token</param>
+        public async Task WriteLongAsync(long v, bool isAsync, CancellationToken ct)
```

Comment:
I'd personally think about removing these and force the callers to build up a well-formed `Span` or `Memory`. IMO we're currently encouraging upstream code to perform writes field by field.

NB: in this case, would we need the class at all? The `TdsStream` handles buffering, we'd make building/writing the on-the-wire structure the responsibility of the caller, are there any other reasons to keep it?

Reply:
Main reasons are to provide encapsulated APIs, and handle buffer allocations. We're keeping byte[] usage as it is at the moment, to keep simplicity, and let Stream write continuously, Something to revisit in future, we should analyse buffer requirements before enforcing that I think.